server/src/main/java/org/elasticsearch/search/retriever/RetrieverBuilder.java

```diff
@@ -0,0 +1,234 @@
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.xcontent.AbstractObjectParser;
+import org.elasticsearch.xcontent.FilterXContentParserWrapper;
+import org.elasticsearch.xcontent.NamedObjectNotFoundException;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentLocation;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * A retriever represents an API element that returns an ordered list of top
+ * documents. These can be obtained from a query, from another retriever, etc.
+ * Internally, a {@link RetrieverBuilder} is just a wrapper for other search
+ * elements that are extracted into a {@link SearchSourceBuilder}. The advantage
+ * retrievers have is in the API they appear as a tree-like structure enabling
+ * easier reasoning about what a search does.
+ *
+ * This is the base class for all other retrievers. This class does not support
+ * serialization and is expected to be fully extracted to a {@link SearchSourceBuilder}
+ * prior to any transport calls.
+ */
+public abstract class RetrieverBuilder implements ToXContent {
+
+    public static final NodeFeature RETRIEVERS_SUPPORTED = new NodeFeature("retrievers_supported");
+
+    public static final ParseField PRE_FILTER_FIELD = new ParseField("filter");
+
+    protected static void declareBaseParserFields(
+        String name,
+        AbstractObjectParser<? extends RetrieverBuilder, RetrieverParserContext> parser
+    ) {
+        parser.declareObjectArray((r, v) -> r.preFilterQueryBuilders = v, (p, c) -> {
+            QueryBuilder preFilterQueryBuilder = AbstractQueryBuilder.parseTopLevelQuery(p, c::trackQueryUsage);
+            c.trackSectionUsage(name + ":" + PRE_FILTER_FIELD.getPreferredName());
+            return preFilterQueryBuilder;
+        }, PRE_FILTER_FIELD);
+    }
+
+    /**
+     * This method parsers a top-level retriever within a search and tracks its own depth. Currently, the
+     * maximum depth allowed is limited to 2 as a compound retriever cannot currently contain another
+     * compound retriever.
+     */
+    public static RetrieverBuilder parseTopLevelRetrieverBuilder(XContentParser parser, RetrieverParserContext context) throws IOException {
+        parser = new FilterXContentParserWrapper(parser) {
+
+            int nestedDepth = 0;
+
+            @Override
+            public <T> T namedObject(Class<T> categoryClass, String name, Object context) throws IOException {
+                if (categoryClass.equals(RetrieverBuilder.class)) {
+                    nestedDepth++;
+
+                    if (nestedDepth > 2) {
```

Comment:
It would be nice to break this out as a constant.

Reply:
Absolutely! This needs a bit of a re-design to work with re-rankers, so I'm going to leave it for now. I think depth will be less important than specific ordering in the end.